### PR TITLE
Simplified KBestHaplotypeFinder by replacing recursion with Dijkstra's algorithm

### LINKED ISSUE #3561: Use Dijkstra's algorithm for finding best haplotypes from assembly graph
@vruano has pointed out that our method of finding the best haplotype paths in an assembly graph is equivalent to Dijkstra's algorithm for finding the shortest path in a directed graph, and that the latter is a much simpler implementation.  [ Do I understand this correctly?].

We could simplify a bunch of code without changing the output of any tools by switching the implementation to Dijkstra's algorithm, which is implemented in jgrapht (our graphs extend this package's graph class) and apache commons.

@vruano has also pointed out that our definition of the best paths may not be optimal, which is a separate issue.

### PR DESCRIPTION
**Pull request by davidbenjamin**
Closes #3561.

@vruano could you review this?  Note that I ended up implementing Dijkstra's algorithm instead of using a library, but it's only a few lines of code.  This PR does not affect the outputs of HC or M2 at all.

Also, @vruano, I recall your misgivings about the current haplotype enumeration (which this preserves):

>However, the current algorithm and the k-dijkstra still would show the same problems in terms of doing a suboptimal selection of haplotypes in terms of their coverage of plausible variation.  I had implemented an alternative that fixed that issue . . . simulate haplotypes based on those same furcation likelihoods and wstop when we have not discovered anything new for a while... the problem of such an approach is to make it deterministic.

Although this PR doesn't do that, it could easily be extended to do so just by running Dijkstra's algorithm until you have the amount of variation you want.  That is, instead of terminating when the Dijkstra priority queue is empty or when we have discovered the maximum number of haplotypes, we could terminate based on some `Predicate<List<KBestHaplotype>>` on the list of haplotypes found so far.  And it's deterministic since Dijkstra's algorithm is greedy.

So basically, it's a nice refactoring for now but it also sets up some worthwhile extensions if we want.

### CODE REVIEW

**Comment by vruano** [comment]
No need to change anything really, go ahead and merge once conflicts are addressed and tests pass.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/KBestHaplotypeFinder.java [hunk 1]
```diff
@@ -69,70 +39,60 @@ public KBestHaplotypeFinder(final SeqGraph graph, final Set<SeqVertex> sources,
...
-    /**
-     * Creates an aggregated recursive finder to try all possible source vertices.
-     *
-     * @return never {@code null}.
-     */
-    private KBestSubHaplotypeFinder createAggregatedFinder() {
-        final List<KBestSubHaplotypeFinder> sourceFinders = new ArrayList<>(sources.size());
-        for (final SeqVertex source : sources) {
-            sourceFinders.add(createVertexFinder(source));
+    public List<KBestHaplotype> findBestHaplotypes(final int maxNumberOfHaplotypes) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/KBestHaplotypeFinder.java]: good point.  Done.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/KBestHaplotypeFinderUnitTest.java [hunk 2]
```diff
@@ -439,12 +383,10 @@ public void testIntraNodeInsertionDeletion() {
         graph.addEdges(() -> new BaseEdge(false, 1), top, alt, bot);
 
         @SuppressWarnings("all")
-        final KBestHaplotypeFinder bestPathFinder = new KBestHaplotypeFinder(graph,top,bot);
-
-        Assert.assertEquals(bestPathFinder.size(), 2);
-
-        final Path<SeqVertex,BaseEdge> refPath = bestPathFinder.get(0).path();
-        final Path<SeqVertex,BaseEdge> altPath = bestPathFinder.get(1).path();
+        final List<KBestHaplotype>bestPaths = new KBestHaplotypeFinder(graph,top,bot).findBestHaplotypes();
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/KBestHaplotypeFinderUnitTest.java]: Fixed.
